Support set flag for component configs

pavolloffay opened the pull request:
Signed-off-by: Pavol Loffay <ploffay@redhat.com>

Creating as a draft to collect feedback. Once we agree I will clean it and add tests.

**Description:** 

This patch adds support for configuring components (receivers, processors, exporters...) via `--set` flag. For example `--set=processors.batch.timeout=12s` or `--receivers.otlp.protocols.grpc.endpoint=123456`.

```
--set stringArray             Set arbitrary component config property. The flag has a higher precedence over config file. The arrays are overridden. Example --set=processors.batch.timeout=2s
```

- Flags have higher precedence than config file
- The component has to be defined in the config to apply the configuration otherwise the flag is noop e.g. `--set=processor.queued_retry.queue_size=15` is noop if the `queued_retry` is not defined.
- The named components work e.g. `--set=processors.batch/bar.timeout=3s`
- The maps are joined with the existing (file) config
- The arrays are overridden 
- The object arrays are more complicated - with flag it's possible to set only the first entry e.g. `--set=processors.resource.attributes.key=key2`


The implementation creates a new viper instance that is initialized with a temporary properties file that contains values passed via set flag. Then the new viper is used to apply changes to an existing configuration (loaded via the main config file). 


**Link to tracking Issue:** <Issue number if applicable>

Resolves #873 

**Testing:** < Describe what testing was performed and which tests were added.>

Tested locally
`RUN_CONFIG=config.yaml make run`

```yaml
receivers:
    otlp:
        protocols:
            grpc:

exporters:
    logging:
        logLevel: debug

processors:
    batch:

service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging]
```

**Documentation:** < Describe the documentation added.>

None

pavolloffay commented:
@tigrannajaryan @jrcamp could you please review and comment whether you agree with this approach?

jrcamp commented:
In `--set=processors.batch.timeout=70s` Are the ='s required? Does `--set processors.batch.timeout=70s` and `--set processors.batch.timeout 70s` work?

pavolloffay commented:
The `s` is required bc it is golang `time.Duration`.

* `--set=processors.batch.timeout=70s` works
* `--set processors.batch.timeout=70s` works after trimming spaces 
* `--set processors.batch.timeout 70s` does not work

pavolloffay commented:
Changing this to ready for review.

pavolloffay commented:
The test coverage is below, there are some parts of this PR that I am not sure how to cover (error of creating temp file, err for writing to a temp file, getting an error from ViperSubExtract)

pavolloffay commented on config/config.go:
```diff
@@ -319,23 +310,6 @@ func loadService(v *viper.Viper) (configmodels.Service, error) {
 }
 
-// LoadReceiver loads a receiver config from componentConfig using the provided factories.
-func LoadReceiver(componentConfig *viper.Viper, typeStr configmodels.Type, fullName string, factory component.ReceiverFactory) (configmodels.Receiver, error) {
```
I wasn't sure why it is public since it is the only public load func for a component.

pavolloffay commented:
> I feel like this can be substantially simplified. See code here based on some of your changes:

> jrcamp@5ea89ed

Maybe I am doing something wrong but it does not work as expected, the set flags are not being enforced. There are also error messages in the log 

```
ERROR 2020/08/31 08:10:37 svType != tvType; key=batch, st=map[string]interface {}, tt=<nil>, sv=map[timeout:12s], tv=<nil>
```

pavolloffay commented:
@jrcamp I made the suggested changes and also the simplification to remove the temp file (great catch!). Your impl does not seem to work properly, the merge config in viper is not working as we want (see my previous comment).

jrcamp commented:
> > I feel like this can be substantially simplified. See code here based on some of your changes:
> 
> > [jrcamp/opentelemetry-collector@5ea89ed](https://github.com/jrcamp/opentelemetry-collector/commit/5ea89ed)
> 
> Maybe I am doing something wrong but it does not work as expected, the set flags are not being enforced. There are also error messages in the log
> 
> ```
> ERROR 2020/08/31 08:10:37 svType != tvType; key=batch, st=map[string]interface {}, tt=<nil>, sv=map[timeout:12s], tv=<nil>
> ```

I only did a quick test so there could be a bug. Can you share the config and command line params you used?

pavolloffay commented:
> I only did a quick test so there could be a bug. Can you share the config and command line params you used?

The config is in the first comment.  Then you can use `--set=processors.batch.timeout=3s` or `--set=exporters.otlp.protocols.grpc.endpoint=localhost:12345`.

jrcamp commented:
@pavolloffay looks like viper `MergeConfigMap` has issues. See latest change https://github.com/jrcamp/opentelemetry-collector/commit/flags-review

pavolloffay commented:
@jrcamp does not seem to work either try the config from the first comment with `--set=exporters.otlp.protocols.grpc.endpoint=localhost:12345`. It fails on 

```
GO111MODULE=on go run --race ./cmd/otelcol/... --config config.yaml --set=exporters.otlp.protocols.grpc.endpoint=localhost:12345
2020/09/04 16:14:12 proto: duplicate proto type registered: opentelemetry.proto.metrics.v1.ResourceMetrics
2020/09/04 16:14:12 proto: duplicate proto type registered: opentelemetry.proto.metrics.v1.InstrumentationLibraryMetrics
2020/09/04 16:14:12 proto: duplicate proto type registered: opentelemetry.proto.metrics.v1.Metric
2020/09/04 16:14:12 proto: duplicate proto type registered: opentelemetry.proto.metrics.v1.DoubleDataPoint
2020-09-04T16:14:12.576+0200	INFO	service/service.go:407	Starting OpenTelemetry Collector...	{"Version": "latest", "GitHash": "<NOT PROPERLY GENERATED>", "NumCPU": 8}
2020-09-04T16:14:12.576+0200	INFO	service/service.go:251	Setting up own telemetry...
2020-09-04T16:14:12.590+0200	INFO	service/telemetry.go:106	Serving Prometheus metrics	{"address": "localhost:8888", "legacy_metrics": false, "new_metrics": true, "level": 3, "service.instance.id": "df297e01-87c1-4ae0-a5c7-72d01f7a9b1c"}
2020-09-04T16:14:12.592+0200	INFO	service/service.go:288	Loading configuration...
Error: cannot load configuration: error reading exporters configuration for otlp: 1 error(s) decoding:

* '' has invalid keys: protocols
2020/09/04 16:14:12 application run finished with error: cannot load configuration: error reading exporters configuration for otlp: 1 error(s) decoding:

* '' has invalid keys: protocols
exit status 1
make: *** [Makefile:163: run] Error 1
```

jrcamp commented:
@pavolloffay I get the same startup error by adding the equivalent set command to the config file. Maybe you meant `--set=receivers.otlp.protocols.grpc.endpoint=localhost:12345` instead?

But there are still more issues even after changing that. See my latest changes (viper isn't very good at merging) https://github.com/jrcamp/opentelemetry-collector/tree/flags-review

I didn't fix the tests for the stuff I refactored, just trying to find an approach that works.

pavolloffay commented:
@jrcamp I have updated the PR with the workaround from your branch and added tests. Now it looks good to me.

```go
for _, k := range v.AllKeys() {
		v.Set(k, v.Get(k))
	}
```

jrcamp commented:
As part of my proposed implementation I also did some minor refactoring so that this isn't tied to `FileLoaderConfigFactory`. In theory this should work for any config factory. Otherwise it's going to be confusing why set flags work in some cases but not in others.

pavolloffay commented:
> As part of my proposed implementation I also did some minor refactoring so that this isn't tied to FileLoaderConfigFactory. 

I saw that the problem is that the factory should return the final config object. We use it to construct the default/hardcoded configuration in Jaeger (agent, collector...).  This remains me that we should make the set flag public.

pavolloffay commented:
> As part of my proposed implementation I also did some minor refactoring so that this isn't tied to FileLoaderConfigFactory. In theory this should work for any config factory. Otherwise it's going to be confusing why set flags work in some cases but not in others.

@jrcamp as I mentioned before the factory has to return the config object. I will document your concerns and make the `applySetFlag` public so factories can use it.

jrcamp commented:
Sounds good. Overall lgtm, can you just note (in the PR here) the reason patch is under target (95%). Looks like it's mostly error cases that would be hard to reach but they may ask before merging.

pavolloffay commented:
Yes, the coverage dropped below the limit. I don't know how to cover error cases e.g.

```go
// b is &bytes.Buffer{}
if _, err := fmt.Fprintf(b, "%s\n", property); err != nil {
                        return err
                }
```

and
```go
if err := viperFlags.ReadConfig(b); err != nil {
                return fmt.Errorf("failed to read set flag config: %v", err)
        }
```

pavolloffay commented:
@tigrannajaryan @bogdandrutu could you please have review/merge?

tigrannajaryan commented:
>The component has to be defined in the config to apply the configuration otherwise the flag is noop e.g. --set=processor.queued_retry.queue_size=15 is noop if the queued_retry is not defined.

Is this a silent noop? Can we make this an error instead of noop? Otherwise it will be difficult to understand why things don't work if you make a typo in the command line option.

tigrannajaryan commented:
> This patch adds support for configuring components (receivers, processors, exporters...) via `--set` flag. For example `--set=processors.batch.timeout=12s` or `--receivers.otlp.protocols.grpc.endpoint=123456`.
> 
> ```
> --set stringArray             Set arbitrary component config property. The flag has a higher precedence over config file. The arrays are overridden. Example --set=processors.batch.timeout=2s
> ```
> 
>     * Flags have higher precedence than config file
> 
>     * The component has to be defined in the config to apply the configuration otherwise the flag is noop e.g. `--set=processor.queued_retry.queue_size=15` is noop if the `queued_retry` is not defined.
> 
>     * The named components work e.g. `--set=processors.batch/bar.timeout=3s`
> 
>     * The maps are joined with the existing (file) config
> 
>     * The arrays are overridden
> 
>     * The object arrays are more complicated - with flag it's possible to set only the first entry e.g. `--set=processors.resource.attributes.key=key2`
> 

Can you add this description to some relevant place in the source code?

We will also need it in the README, but it seems like we lost the basic running instructions after refactoring, so this can wait for now.

I just want to make sure the usage instructions are captured at least somewhere. Perhaps we can print them when `--help` is invoked?

pavolloffay commented on service/service_test.go:
```diff
@@ -511,6 +518,105 @@ func TestApplication_GetExporters(t *testing.T) {
+	params := Parameters{
+		Factories: factories,
+	}
+	t.Run("unknown_component", func(t *testing.T) {
```
@tigrannajaryan have a look at this test.

If the flag references an unknown property the factory returns an error. However, if the flag references a valid component that is not in the config file e.g. `batch/foo` then an error is not returned (see the next test) and a new component instance is created that is not enabled in the pipeline. I think that this behaviour is fine.

tigrannajaryan replied:
>However, if the flag references a valid component that is not in the config file e.g. batch/foo then an error is not returned (see the next test) and a new component instance is created that is not enabled in the pipeline. I

This is the part that I don't like. If I have a typo in my `--set` flag and the name of the component is incorrect then it will have no effect and there will be no indication to the user that something went wrong.

Preferably in this case the startup should fail and tell that the component is not found. Is it possible to do?

pavolloffay replied:
I cannot think of a simple solution. The same can happen with the config file a user can forget to add a component to a pipeline.

The ugly solution might be to parse the config in `FileLoaderConfigFactory` then pass it into `AddSetFlagProperties` which would parse it's own config and compare which components are not in the pipeline. Then parse the final config at the end of `FileLoaderConfigFactory` with the final viper. This is a rather horrible solution :).

Maybe we could print a warning for not assigned components during the config validation?

tigrannajaryan commented on service/service.go:
```diff
@@ -104,17 +104,19 @@ type Parameters struct {
 	// If it is not provided the default factory (FileLoaderConfigFactory) is used.
-	// The default factory loads the configuration specified as a command line flag.
+	// The default factory loads the configuration file and overrides component's configuration
+	// properties supplied via --set flag..
```
```suggestion
	// properties supplied via --set command line flag.
```

tigrannajaryan commented on service/service.go:
```diff
@@ -104,17 +104,19 @@ type Parameters struct {
+// The ConfigFactory implementation should call AddSetFlagProperties to enable configuration passed via `--set` flag.
+type ConfigFactory func(v *viper.Viper, cmd *cobra.Command, factories component.Factories) (*configmodels.Config, error)
 
 // FileLoaderConfigFactory implements ConfigFactory and it creates configuration from file.
```
```suggestion
// FileLoaderConfigFactory implements ConfigFactory and it creates configuration from file
// and from --set command line flag (if the flag is present).
```

tigrannajaryan commented on service/service.go:
```diff
@@ -104,17 +104,19 @@ type Parameters struct {
 // ConfigFactory creates config.
-type ConfigFactory func(v *viper.Viper, factories component.Factories) (*configmodels.Config, error)
+// The ConfigFactory implementation should call AddSetFlagProperties to enable configuration passed via `--set` flag.
+type ConfigFactory func(v *viper.Viper, cmd *cobra.Command, factories component.Factories) (*configmodels.Config, error)
```
Please add description of parameters.

tigrannajaryan commented on service/service.go:
```diff
@@ -124,6 +126,14 @@ func FileLoaderConfigFactory(v *viper.Viper, factories component.Factories) (*co
 	if err != nil {
 		return nil, fmt.Errorf("error loading config file %q: %v", file, err)
 	}
+	for _, k := range v.AllKeys() {
+		v.Set(k, v.Get(k))
+	}
```
Why do we do this? Code is not clear and looks suspect (we get and set all keys?). Add comment.

pavolloffay replied:
I have added a comment

tigrannajaryan commented on service/service_test.go:
```diff
@@ -38,6 +41,10 @@ import (
 	"go.opentelemetry.io/collector/component/componenttest"
 	"go.opentelemetry.io/collector/config"
 	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/processor/attributesprocessor"
+	"go.opentelemetry.io/collector/processor/batchprocessor"
+	"go.opentelemetry.io/collector/receiver/jaegerreceiver"
+	"go.opentelemetry.io/collector/service/builder"
```
I am not sure I like how we depend on specific components from here. Can we use `componenttest` and example components for testing instead?

I would prefer to minimize the coupling between this package and actual component packages.

pavolloffay replied:
The test was already using `"go.opentelemetry.io/collector/service/defaultcomponents"`. The components were parsed but not accessed directly. For instance this config was used by this test https://github.com/open-telemetry/opentelemetry-collector/blob/9778b16b81016d6035b2a9f42cda8dc05d8165b8/service/testdata/otelcol-config.yaml#L1

tigrannajaryan commented on service/set_flag.go:
```diff
@@ -0,0 +1,66 @@
+		return fmt.Errorf("failed to read set flag config: %v", err)
+	}
+
+	for _, k := range viperFlags.AllKeys() {
+		v.Set(k, viperFlags.Get(k))
+	}
```
Why do we build a properties file first? Can't we apply the `flagProperties` directly to `v` here? If there is a reason please add comments to explain.

pavolloffay replied:
I have added a comment to the source file.

pavolloffay commented:
> I just want to make sure the usage instructions are captured at least somewhere. Perhaps we can print them when --help is invoked?

This is what the help command prints:

```
--set stringArray             Set arbitrary component config property. The component has to be defined in the config file and the flag has a higher precedence. Array config properties are overridden and maps are joined, note that only a single (first) array property can be set e.g. -set=processors.attributes.actions.key=some_key. Example --set=processors.batch.timeout=2s
```

> The component has to be defined in the config to apply the configuration otherwise the flag is noop e.g. --set=processor.queued_retry.queue_size=15 is noop if the queued_retry is not defined.

> Is this a silent noop? Can we make this an error instead of noop? Otherwise it will be difficult to understand why things don't work if you make a typo in the command line option.

See the help command it says that the component has to be defined in the config file. I am not sure how to do this, nicely. We could iterate over components and see whether they are defined in a pipeline. The same issue is with a config file a user can forget to add a component to the pipeline.

pavolloffay commented:
@tigrannajaryan could you please have another look?

pavolloffay commented:
@tigrannajaryan any update on this one?
